### DISCUSSION TITLE
Namespace proliferation

### DISCUSSION THREAD

**Discussion by AshleyRoll** [intel/compile-time-init-build#299] [General]
I'm relatively new to this library and I'm aware the library itself is new  and under construction, so I wanted to raise the question of namespaces for the library as it seems there is currently one namespace per module. Perhaps this stems from the all of these pieces being combined from other sources into a single library, but I'm not certain of the history here from outside the main developer group.

As a potential consumer of the library I would expect that it introduces a single top level namespace, however I'm not sure what others would like here.

We currently have the following top level namespaces (at a rough glance):

- `cib`
- `flow`
- `interrupt`
- `logging`
- `msg`
- `sc`
- `seq`

There are also containers living in the global namespace see #298 

I would like to propose we use the single global namespace `cib` and move the others below this.

This would likely introduce some minor rework in existing code, but I believe it would make the library easier for existing code bases or new ones to adopt. `logging` and `msg` and perhaps `interrupt` might already be common namespaces in existing code (I know I have 2 of them inside an embedded project).

**Comment by lukevalenty** [comment]
That's a very good point. The proliferation is 100% on me. I got lazy and started putting closely related libraries into this repo rather than releasing them as their own.

There should definitely be a different top-level namespace for all of these. I'm OK with `cib`, but should discuss other options. If we want to pull some of these libraries out of the cib repo, then a different top-level namespace may be more appropriate.

**Comment by AshleyRoll** [comment]
I don't mind that all the libraries are combined into a single tool. Its not like you would likely just use one. Splitting into finer grain libraries just moves the problem to one of dependency management, and you still have potentially conflicting namespaces.

I don't mind `cib` being top level. Do we leave the current `cib` code at that level or rename it something like `cib::nexus`?

Conscious that there is existing code using this that will be impacted.
